### PR TITLE
Roomoval

### LINKED ISSUE #705: The Roomoval
We want to switch to Rust for migrations (#704). To do that, we need to stop using Room, which is very opinionated about how migrations are handled.

### PR DESCRIPTION
**Pull request by ccjernigan**
Closes #705.

The Android and Rust code have previously managed joint custody of the derived data database.  With more complex migrations now required, we need to make the Android side read-only.  To achieve that, the Android side will remove Room and instead rely on more primitive SQLite APIs for read-only access.

As part of implementing this change, database management throughout the SDK is being refactored.

There will be multiple representations of the data:
 - Database representation (and some Entity representations in the places that Room hasn’t been removed yet).  These representations are not as type safe and don’t match Kotlin best practices in all cases.
 - Once #615 is implemented there will also be network representations
 - Type safe models, which often appear in the public API  The Database and Network representations will be converted to and from the type safe model representation.

Still pending before merging:
 - Many database methods are stubbed out.  See DbDerivedDataRepository
 - Some database write operations are occurring and these should be replaced with Rust implementations.  Many of these are happening inside SdkSynchronizer
 - Unit testing of the Android database query implementations (given the Rust layer has been set up and known query results are expected back out)

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/tests) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [ ] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the demo app and try the changes?
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the demo app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/db/derived/TransactionTable.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.withContext
+import java.util.Locale
+
+internal class TransactionTable(private val zcashNetwork: ZcashNetwork, private val sqliteDatabase: SQLiteDatabase) {
+
+    companion object {
+        private val SELECTION_BLOCK_IS_NULL = String.format(
```
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/db/derived/TransactionTable.kt:18]: Formatting: Some `companion object` definitions have a blank line interleaved with the first statements and others don't. Which of the two is the convention?

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/Account.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package cash.z.ecc.android.sdk.internal.model
+
+internal data class Account(
+    val account: Int,
+    val extendedFullViewingKey: String,
+    val address: String,
+    val transparentAddress: String
+) {
+    // Override to prevent leaking details in logs
+    override fun toString() = "Account"
```
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/Account.kt:10]: 👍

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/PendingTransaction.kt [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    // TODO: test for off-by-one error here. Should we use <= or <
+    if (latestHeight.value < saplingActivationHeight.value || expiryHeightLocal < saplingActivationHeight) {
+        return false
+    }
+
+    return expiryHeightLocal < latestHeight
+}
+
+private const val EXPIRY_BLOCK_COUNT = 100
```
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/PendingTransaction.kt:88]: This value is probably defined as a constant on the SDK already

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/PendingTransaction.kt [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+private const val ERROR_CODE_MARKED_FOR_DELETION = -9090
+
+fun PendingTransaction.isMarkedForDeletion(): Boolean {
+    return rawTransactionId == null && (errorCode ?: 0) == ERROR_CODE_MARKED_FOR_DELETION
+}
+
+private val smallThreshold = 30.minutes
+private val hugeThreshold = 30.days
+
+fun PendingTransaction.isSafeToDiscard(): Boolean {
```
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/model/PendingTransaction.kt:117]: I know this maintains current functionality. But it should be important to revisit deletion of submission failures without consent of the user.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package cash.z.ecc.android.sdk.model
+
+data class Received(
```
Comment (pacu) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt:3]: Shouldn't these be called ReceivedNote and SentNote? They could be mistakenly thought as transactions

**Review by pacu — COMMENTED** [review]
First pass to review this PR. Left some comments and remark nothing blocking

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt (hunk 5 above)
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt:3]: I will rename them 👍

**Comment by ccjernigan** [comment]
Closing this PR in favor of a branch targeting the long-lived feature branch https://github.com/zcash/zcash-android-wallet-sdk/tree/feature/roomoval

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt (hunk 5 above)
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/Received.kt:3]: Actually we went backwards on this and made them more clearly Transactions and not notes.

**Pull request closed without merging**